src/main/java/tlschannel/async/AsynchronousTlsChannelGroup.java

```diff
@@ -369,8 +365,11 @@ private void waitForSocketRegistration(RegisteredSocket socket) {
   private void loop() {
     try {
       while (shutdown == Shutdown.No
-          || shutdown == Shutdown.Wait && currentRegistrations.intValue() > 0) {
-        int c = selector.select(); // block
+          || shutdown == Shutdown.Wait
+              && (!pendingRegistrations.isEmpty() || !registrations.isEmpty())) {
+        // most state-changing operations will wake the selector up, however, asynchronous closings
+        // of the channels won't, so we have to timeout to allow checking those cases
+        int c = selector.select(100); // block
```

Comment:
I am writing this question in order to clarify my understanding.

This timeout comes into play only if `RegisteredSocket.socketChannel` is closed by a user directly, as opposed to be closed via the `AsynchronousTlsChannel.close` method, right? If the socket is closed via `AsynchronousTlsChannel.close`, then `selector.wakeup` is called, which causes `selector.select` to return ASAP. In other words, users that use `AsynchronousTlsChannel.close` will not experience 100 ms delays between calling `AsynchronousTlsChannel.close` and having a read/write operation failed with `ClosedChannelException`.

Reply:
> This timeout comes into play only if RegisteredSocket.socketChannel is closed by a user directly, as opposed to be closed via the AsynchronousTlsChannel.close method, right?

Right.

> If the socket is closed via AsynchronousTlsChannel.close, then selector.wakeup is called, which causes selector.select to return ASAP. In other words, users that use AsynchronousTlsChannel.close will not experience 100 ms delays between calling AsynchronousTlsChannel.close and having a read/write operation failed with ClosedChannelException.

Exactly. it's only for "out of band" closings. And the value obviously a compromise between responsiveness in this not-so-common use case and CPU usage.

> While previously there was no timeout at all, RegisteredSocket.close was explicitly cancelling read/write operations. This is no longer the case, hence the question.

Yes, I removed the explicit cancellation because I realized that a close() and a Selector.wakeup() would be essentially the same now that we are checking that in every iteration of the loop.

Reply:
Thank you for implementing the fix and for the answers. I would not even bother working around situations when `AsynchronousTlsChannel.close` is not called: it's a `Closeable`, and users must close such objects. If that is not done, then all bets are off.

I don't see problems with the changes to `AsynchronousTlsChannelGroup`, and would be happy to see them merged, so that we could adopt them to the mongo-java-driver.